airflow/cli/api/client.py

```diff
@@ -0,0 +1,228 @@
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import base64
+import json
+import os
+import sys
+from typing import TYPE_CHECKING, Any
+
+import httpx
+import rich
+import structlog
+from uuid6 import uuid7
+
+from airflow.cli.api.operations import (
+    AssetsOperations,
+    BackfillsOperations,
+    ConfigOperations,
+    ConnectionsOperations,
+    DagOperations,
+    DagRunOperations,
+    JobsOperations,
+    PoolsOperations,
+    ProvidersOperations,
+    ServerResponseError,
+    VariablesOperations,
+    VersionOperations,
+)
+from airflow.utils.types import DagRunType
+from airflow.version import version
+
+if TYPE_CHECKING:
+    # # methodtools doesn't have typestubs, so give a stub
+    def lru_cache(maxsize: int | None = 128):
+        def wrapper(f):
+            return f
+
+        return wrapper
+else:
+    from methodtools import lru_cache
+
+log = structlog.get_logger(logger_name=__name__)
+
+__all__ = [
+    "Client",
+    "Credentials",
+]
+
+
+def add_correlation_id(request: httpx.Request):
+    request.headers["correlation-id"] = str(uuid7())
+
+
+def get_json_error(response: httpx.Response):
+    """Raise a ServerResponseError if we can extract error info from the error."""
+    err = ServerResponseError.from_response(response)
+    if err:
+        log.warning("Server error ", extra=dict(err.response.json()))
+        raise err
+
+
+def raise_on_4xx_5xx(response: httpx.Response):
+    return get_json_error(response) or response.raise_for_status()
+
+
+def noop_handler(request: httpx.Request) -> httpx.Response:
+    path = request.url.path
+    log.debug("Dry-run request", method=request.method, path=path)
+    # TODO change for test
+    if path.startswith("/task-instances/") and path.endswith("/run"):
+        # Return a fake context
+        return httpx.Response(
+            200,
+            json={
+                "dag_run": {
+                    "dag_id": "test_dag",
+                    "run_id": "test_run",
+                    "logical_date": "2021-01-01T00:00:00Z",
+                    "start_date": "2021-01-01T00:00:00Z",
+                    "run_type": DagRunType.MANUAL,
+                },
+            },
+        )
+    return httpx.Response(200, json={"text": "Hello, world!"})
+
+
+# Credentials for the API
+class Credentials:
+    """Credentials for the API."""
+
+    api_url: str
+    api_token: str
+
+    def __init__(self, api_url: str, api_token: str = "NO_TOKEN"):
+        self.api_url = api_url
+        self.api_token = api_token
+
+    def save(self):
+        """Save the credentials."""
+        default_config_dir = os.path.expanduser("~/.config/airflow")
```

Comment:
Please respect [`$XDG_CONFIG_HOME`](https://specifications.freedesktop.org/basedir-spec/latest/) if we are going to put files on disk.

Reply:
(see my comment in doc about using https://pypi.org/project/keyring/ instead though)

Reply:
Both [$XDG_CONFIG_HOME](https://specifications.freedesktop.org/basedir-spec/latest/) and keyring are integrated. Thanks!